{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db61e737",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "In this notebook, we will walk through an example of using our pairwise regression model to predict RNA-seq in LCLs on the test individuals for the EFTUD2 gene. \n",
    "\n",
    "Along the way, we will detail three important artifacts of our study.\n",
    "- the pre-processed GEUVADIS RNA-seq data\n",
    "- a h5 object containing one-hot encoded sequences of GEUVADIS individuals\n",
    "- the pairwise regression model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9ba01e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "REPO_DIR = \"/oak/stanford/groups/akundaje/rrastogi/external_repos/finetuning-enformer\" # Path to the finetuning-enformer repo\n",
    "DOWNLOAD_DIR = \".download\" # Directory to store downloaded data (sequences, models, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "88c1b6be",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import h5py\n",
    "import os\n",
    "import sys\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import torch\n",
    "from scipy.stats import pearsonr\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "# In case you don't have finetuning-enformer installed as a package, add it to the system path\n",
    "sys.path.append(os.path.join(REPO_DIR, \"finetuning\"))\n",
    "from models import PairwiseRegressionFloatPrecision"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55a6135d",
   "metadata": {},
   "source": [
    "# 1. Load true (processed) gene expression data for EFTUD2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dcf3c289",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXPRESSION_DATA_PATH = os.path.join(REPO_DIR, \"process_geuvadis_data\", \"log_tpm\", \"corrected_log_tpm.annot.csv.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "99f595fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_data = pd.read_csv(EXPRESSION_DATA_PATH, index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5998ccb",
   "metadata": {},
   "source": [
    "The processed expression dataframe contains log TPM values that have been corrected for batch effects/population stratification by regressing out the top 10 expression PCs.\n",
    "\n",
    "Each row corresponds to a different gene. Most columns refer to GEUVADIS donor IDs; other columns contain gene metadata, such  as whether a significant eQTL was detected in the European or Yoruba population in the original Geuvadis analysis.\n",
    "\n",
    "Note: we only train and evaluate on genes where a significant eQTL was detected in the European population, but this table contains information for all genes that are not\n",
    "lowly expressed in LCLs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cefd6186",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Coord</th>\n",
       "      <th>HG00096</th>\n",
       "      <th>HG00097</th>\n",
       "      <th>HG00099</th>\n",
       "      <th>HG00100</th>\n",
       "      <th>HG00101</th>\n",
       "      <th>HG00102</th>\n",
       "      <th>HG00103</th>\n",
       "      <th>...</th>\n",
       "      <th>NA20828</th>\n",
       "      <th>stable_id</th>\n",
       "      <th>gencode_v12_gene_name</th>\n",
       "      <th>our_gene_name</th>\n",
       "      <th>EUR_eGene</th>\n",
       "      <th>YRI_eGene</th>\n",
       "      <th>top_EUR_eqtl_rsid</th>\n",
       "      <th>top_YRI_eqtl_rsid</th>\n",
       "      <th>top_EUR_eqtl_distance</th>\n",
       "      <th>top_YRI_eqtl_distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TargetID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000257527.1</th>\n",
       "      <td>ENSG00000257527.1</td>\n",
       "      <td>16</td>\n",
       "      <td>18505708</td>\n",
       "      <td>-0.057361</td>\n",
       "      <td>-0.313160</td>\n",
       "      <td>-0.684395</td>\n",
       "      <td>-1.209085</td>\n",
       "      <td>-0.012644</td>\n",
       "      <td>-0.270612</td>\n",
       "      <td>-0.930251</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.127696</td>\n",
       "      <td>ENSG00000257527</td>\n",
       "      <td>rp11-1212a22.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000151503.7</th>\n",
       "      <td>ENSG00000151503.7</td>\n",
       "      <td>11</td>\n",
       "      <td>134095348</td>\n",
       "      <td>3.653703</td>\n",
       "      <td>3.555238</td>\n",
       "      <td>3.969966</td>\n",
       "      <td>3.832266</td>\n",
       "      <td>3.620463</td>\n",
       "      <td>3.682108</td>\n",
       "      <td>3.862410</td>\n",
       "      <td>...</td>\n",
       "      <td>3.984807</td>\n",
       "      <td>ENSG00000151503</td>\n",
       "      <td>ncapd3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000254681.2</th>\n",
       "      <td>ENSG00000254681.2</td>\n",
       "      <td>16</td>\n",
       "      <td>18495797</td>\n",
       "      <td>2.088882</td>\n",
       "      <td>2.326419</td>\n",
       "      <td>2.128807</td>\n",
       "      <td>2.199625</td>\n",
       "      <td>2.331783</td>\n",
       "      <td>2.627187</td>\n",
       "      <td>1.608311</td>\n",
       "      <td>...</td>\n",
       "      <td>1.565265</td>\n",
       "      <td>ENSG00000254681</td>\n",
       "      <td>rp11-1212a22.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228477.1</th>\n",
       "      <td>ENSG00000228477.1</td>\n",
       "      <td>1</td>\n",
       "      <td>40428352</td>\n",
       "      <td>5.579332</td>\n",
       "      <td>5.352685</td>\n",
       "      <td>5.758683</td>\n",
       "      <td>6.045576</td>\n",
       "      <td>5.563191</td>\n",
       "      <td>5.176924</td>\n",
       "      <td>5.579479</td>\n",
       "      <td>...</td>\n",
       "      <td>5.187391</td>\n",
       "      <td>ENSG00000228477</td>\n",
       "      <td>rp3-342p20.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000159733.9</th>\n",
       "      <td>ENSG00000159733.9</td>\n",
       "      <td>4</td>\n",
       "      <td>2420390</td>\n",
       "      <td>-0.984586</td>\n",
       "      <td>-1.124469</td>\n",
       "      <td>-0.433654</td>\n",
       "      <td>-1.025796</td>\n",
       "      <td>-0.705150</td>\n",
       "      <td>-1.333362</td>\n",
       "      <td>-0.532541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.044033</td>\n",
       "      <td>ENSG00000159733</td>\n",
       "      <td>zfyve28</td>\n",
       "      <td>zfyve28</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>rs4974687</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9347.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 474 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Gene_Symbol Chr      Coord   HG00096   HG00097  \\\n",
       "TargetID                                                                  \n",
       "ENSG00000257527.1  ENSG00000257527.1  16   18505708 -0.057361 -0.313160   \n",
       "ENSG00000151503.7  ENSG00000151503.7  11  134095348  3.653703  3.555238   \n",
       "ENSG00000254681.2  ENSG00000254681.2  16   18495797  2.088882  2.326419   \n",
       "ENSG00000228477.1  ENSG00000228477.1   1   40428352  5.579332  5.352685   \n",
       "ENSG00000159733.9  ENSG00000159733.9   4    2420390 -0.984586 -1.124469   \n",
       "\n",
       "                    HG00099   HG00100   HG00101   HG00102   HG00103  ...  \\\n",
       "TargetID                                                             ...   \n",
       "ENSG00000257527.1 -0.684395 -1.209085 -0.012644 -0.270612 -0.930251  ...   \n",
       "ENSG00000151503.7  3.969966  3.832266  3.620463  3.682108  3.862410  ...   \n",
       "ENSG00000254681.2  2.128807  2.199625  2.331783  2.627187  1.608311  ...   \n",
       "ENSG00000228477.1  5.758683  6.045576  5.563191  5.176924  5.579479  ...   \n",
       "ENSG00000159733.9 -0.433654 -1.025796 -0.705150 -1.333362 -0.532541  ...   \n",
       "\n",
       "                    NA20828        stable_id  gencode_v12_gene_name  \\\n",
       "TargetID                                                              \n",
       "ENSG00000257527.1 -1.127696  ENSG00000257527         rp11-1212a22.6   \n",
       "ENSG00000151503.7  3.984807  ENSG00000151503                 ncapd3   \n",
       "ENSG00000254681.2  1.565265  ENSG00000254681         rp11-1212a22.3   \n",
       "ENSG00000228477.1  5.187391  ENSG00000228477           rp3-342p20.2   \n",
       "ENSG00000159733.9  0.044033  ENSG00000159733                zfyve28   \n",
       "\n",
       "                   our_gene_name  EUR_eGene  YRI_eGene  top_EUR_eqtl_rsid  \\\n",
       "TargetID                                                                    \n",
       "ENSG00000257527.1            NaN      False      False                NaN   \n",
       "ENSG00000151503.7            NaN      False      False                NaN   \n",
       "ENSG00000254681.2            NaN      False      False                NaN   \n",
       "ENSG00000228477.1            NaN      False      False                NaN   \n",
       "ENSG00000159733.9        zfyve28       True      False          rs4974687   \n",
       "\n",
       "                   top_YRI_eqtl_rsid  top_EUR_eqtl_distance  \\\n",
       "TargetID                                                      \n",
       "ENSG00000257527.1                NaN                    NaN   \n",
       "ENSG00000151503.7                NaN                    NaN   \n",
       "ENSG00000254681.2                NaN                    NaN   \n",
       "ENSG00000228477.1                NaN                    NaN   \n",
       "ENSG00000159733.9                NaN                 9347.0   \n",
       "\n",
       "                   top_YRI_eqtl_distance  \n",
       "TargetID                                  \n",
       "ENSG00000257527.1                    NaN  \n",
       "ENSG00000151503.7                    NaN  \n",
       "ENSG00000254681.2                    NaN  \n",
       "ENSG00000228477.1                    NaN  \n",
       "ENSG00000159733.9                    NaN  \n",
       "\n",
       "[5 rows x 474 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "682827fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of donors: 462\n",
      "Example true expression values for EFTUD2:\n",
      "HG00096: 3.5083597317592243\n",
      "HG00097: 3.528019606030445\n",
      "HG00099: 3.27850089238909\n",
      "HG00100: 3.4304214030439257\n",
      "HG00101: 3.53166263223106\n"
     ]
    }
   ],
   "source": [
    "# All donor columns start with either HG or NA\n",
    "donor_columns = [c for c in expression_data.columns if c.startswith(\"HG\") or c.startswith(\"NA\")]\n",
    "print(f\"Number of donors: {len(donor_columns)}\")\n",
    "\n",
    "# Restrict to the EFTUD2 gene and then store values in a dictionary\n",
    "eftud2_data = expression_data[expression_data[\"our_gene_name\"] == \"eftud2\"].iloc[0] # only one row, so select first\n",
    "eftud2_data = eftud2_data[donor_columns]\n",
    "eftud2_true_values = eftud2_data.to_dict()\n",
    "\n",
    "# Display a subset of the true values\n",
    "print(\"Example true expression values for EFTUD2:\")\n",
    "for donor, value in list(eftud2_true_values.items())[:5]:\n",
    "    print(f\"{donor}: {value}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac3daa43",
   "metadata": {},
   "source": [
    "# 2. Load one-hot sequences for individuals in the test set for EFTUD2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8383e4eb",
   "metadata": {},
   "source": [
    "The one-hot encoded sequences for the train, val, and test split have been uploaded to [HuggingFace](https://huggingface.co/anikethjr/finetuning-enformer/tree/main/data/). We will download the test.h5.gz file to our local `DOWNLOAD_DIR` and then extract the one-hot encoded sequences for EFTUD2.\n",
    "\n",
    "*Note*: All the test sequences for random-split genes and population-split genes are in `test.h5.gz`. That file also contains the test sequences for 100 unseen genes. The test sequences for the remaining unseen genes are present in `rest_unseen_filtered.h5.gz`. Because EFTUD2 is a random-split gene, we will just download `test.h5.gz`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9dc832f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "HF_DATA_DIR = \"https://huggingface.co/anikethjr/finetuning-enformer/resolve/main/data\"\n",
    "HF_TEST_H5_PATH = f\"{HF_DATA_DIR}/test.h5.gz?download=true\"\n",
    "\n",
    "LOCAL_TEST_H5_PATH = os.path.join(DOWNLOAD_DIR, \"test.h5.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d211f9a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(DOWNLOAD_DIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c581fdc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  1356  100  1356    0     0   8521      0 --:--:-- --:--:-- --:--:--  8528\n",
      "100 1493M  100 1493M    0     0  77.5M      0  0:00:19  0:00:19 --:--:-- 79.4M   0  0:00:19  0:00:08  0:00:11 79.4M\n"
     ]
    }
   ],
   "source": [
    "# Download can take multiple minutes\n",
    "!curl -L {HF_TEST_PATH} -o {LOCAL_TEST_H5_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7e8a9fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Unzipping can also take several minutes (requires more than 25 GB of disk space)\n",
    "!pigz -d -p 8 {os.path.join(DOWNLOAD_DIR, \"test.h5.gz\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aefbac30",
   "metadata": {},
   "source": [
    "This h5 file contains multiple fields:\n",
    "- `seqs`: one-hot-encoded sequences\n",
    "    - shape: (n_seqs, n_haplotypes = 2, sequence_length = 49152, alphabet_size = 4)\n",
    "- `genes`: gene associated with each sequence\n",
    "    - shape: (n_seqs)\n",
    "- `samples`: donor (individual) associated with each sequence\n",
    "    - shape (n_seqs)\n",
    "- `ancestries`: ancesetry of the donor for each sequence\n",
    "    - shape (n_seqs)\n",
    "- `Y`: log TPM values corrected for top 10 expression PCs\n",
    "    - shape (n_seqs)\n",
    "- `Z`: per-gene z-scores of `Y`, computed across all donors\n",
    "    - shape (n_seqs)\n",
    "- `P`: per-gene percentiles of `Y`, computed across all donors\n",
    "    - shape (n_seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0692dbcb",
   "metadata": {},
   "source": [
    "Here, we will only extract the sequences for individuals in the test set of EFTUD2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# of EFTUD2 individuals in test set: 77\n"
     ]
    }
   ],
   "source": [
    "with h5py.File(LOCAL_TEST_H5_PATH.replace(\".gz\", \"\"), \"r\") as f:\n",
    "    all_genes = f[\"genes\"][:].astype(str)\n",
    "    eftud2_idxs = np.where(all_genes == \"eftud2\")[0]\n",
    "    eftud2_donors = f[\"samples\"][eftud2_idxs].astype(str)\n",
    "    eftud2_seqs = f[\"seqs\"][eftud2_idxs]\n",
    "    eftud2_donor_to_seq = {donor: seq for donor, seq in zip(eftud2_donors, eftud2_seqs)}\n",
    "\n",
    "print(f\"# of EFTUD2 individuals in test set: {len(eftud2_donor_to_seq)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "400c6697",
   "metadata": {},
   "source": [
    "# 3. Make predictions using the pairwise regression model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72538f4a",
   "metadata": {},
   "source": [
    "The weights of all primary models (i.e. those not trained for ablation analyses) are uploaded to [HuggingFace](https://huggingface.co/anikethjr/finetuning-enformer/tree/main/saved_models). Each primary model has three replicates, corresponding to different random seeds. We will first download the seed 42 replicate of the pairwise regression model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "1f1a1015",
   "metadata": {},
   "outputs": [],
   "source": [
    "HF_MODEL_DIR = \"https://huggingface.co/anikethjr/finetuning-enformer/resolve/main/saved_models\"\n",
    "\n",
    "HF_CKPT_PATH = f\"{HF_MODEL_DIR}/regression_data_seed_42_lr_0.0001_wd_0.001_rcprob_0.5_rsmax_3/checkpoints/best.ckpt?download=true\"\n",
    "LOCAL_CKPT_PATH = os.path.join(DOWNLOAD_DIR, \"pairwise_regression_seed_42.ckpt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cad42ba9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  1307  100  1307    0     0   4070      0 --:--:-- --:--:-- --:--:--  4084\n",
      "100 2712M  100 2712M    0     0  69.5M      0  0:00:38  0:00:38 --:--:-- 44.5M\n"
     ]
    }
   ],
   "source": [
    "# Download may take several minutes\n",
    "!curl -L {HF_CKPT_PATH} -o {LOCAL_CKPT_PATH}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0285ea2",
   "metadata": {},
   "source": [
    "Load model onto GPU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7acc932",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert torch.cuda.is_available()\n",
    "device = torch.device(\"cuda:0\")\n",
    "model = PairwiseRegressionFloatPrecision.load_from_checkpoint(LOCAL_CKPT_PATH)\n",
    "model = model.to(device)\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4da9c2c6",
   "metadata": {},
   "source": [
    "Make predictions for each individual by averaging over the forward and reverse complement sequences for both haplotypes.\n",
    "\n",
    "*Note*: The pairwise regression model is trained to predict $z$-scored expression values. This means expression values are only comparable across individuals for a given gene, but not across different genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "987be310",
   "metadata": {},
   "outputs": [],
   "source": [
    "@torch.inference_mode()\n",
    "def make_predictions_for_donor(\n",
    "    model: PairwiseRegressionFloatPrecision,\n",
    "    device: torch.device,\n",
    "    donor_to_seq: dict[str, np.ndarray],\n",
    "    donor: str,\n",
    "):\n",
    "    fwd_seq = donor_to_seq[donor].astype(np.float32) # [n_haplotypes=2, sequence_length=49152, 4]\n",
    "    rc_seq = np.flip(fwd_seq, axis=(-1, -2)).copy()\n",
    "    batch_seq = np.stack([fwd_seq, rc_seq], axis=0) # [n_strands=2, n_haplotypes=2, sequence_length=49152, 4]\n",
    "    \n",
    "    batch_seq_tensor = torch.from_numpy(batch_seq).to(device)\n",
    "    preds = model(batch_seq_tensor) # [n_strands=2]\n",
    "    preds = preds.detach().cpu().numpy()\n",
    "    return np.mean(preds) # average over strands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1fb9170a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2e94ae7f27dd49199d31685f9c8aebbe",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/77 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "eftud2_pred_values = {}\n",
    "for donor in tqdm(eftud2_donor_to_seq):\n",
    "    eftud2_pred_values[donor] = make_predictions_for_donor(\n",
    "        model=model,\n",
    "        device=device,\n",
    "        donor_to_seq=eftud2_donor_to_seq,\n",
    "        donor=donor,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "700209ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_pearson_corr(true_values: dict[str, float], pred_values: dict[str, float]) -> float:\n",
    "    assert set(pred_values.keys()).issubset(set(true_values.keys()))\n",
    "    common_donors = list(pred_values.keys())\n",
    "    x = np.asarray([true_values[donor] for donor in common_donors]) \n",
    "    y = np.asarray([pred_values[donor] for donor in common_donors])\n",
    "    return pearsonr(x, y)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1c805e95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(0.34543399621640836)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compute_pearson_corr(eftud2_true_values, eftud2_pred_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bb3d0dd",
   "metadata": {},
   "source": [
    "Ensure that this matches the performance reported in the `all_gene_perf.csv` file up to a rounding error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "265a1f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_GENE_PERF_PATH = os.path.join(REPO_DIR, \"analysis\", \"all_gene_perf.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d9d8293b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Pearson</th>\n",
       "      <th>|Pearson|</th>\n",
       "      <th>model</th>\n",
       "      <th>class</th>\n",
       "      <th>Chr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>eftud2</td>\n",
       "      <td>0.345252</td>\n",
       "      <td>0.345252</td>\n",
       "      <td>regression_data_seed_42_lr_0.0001_wd_0.001_rcp...</td>\n",
       "      <td>random_split</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene   Pearson  |Pearson|  \\\n",
       "114  eftud2  0.345252   0.345252   \n",
       "\n",
       "                                                 model         class  Chr  \n",
       "114  regression_data_seed_42_lr_0.0001_wd_0.001_rcp...  random_split   17  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_gene_perf_df = pd.read_csv(ALL_GENE_PERF_PATH)\n",
    "all_gene_perf_df[\n",
    "    (all_gene_perf_df[\"gene\"] == \"eftud2\")\n",
    "    & (all_gene_perf_df[\"model\"] == \"regression_data_seed_42_lr_0.0001_wd_0.001_rcprob_0.5_rsmax_3\")\n",
    "]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "multiome_linker",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
