{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74f0c402",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c90a5550",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = \"/clusterfs/nilah/ruchir/src/finetuning-enformer/finetuning/data/h5_bins_384_chrom_split/\"\n",
    "enformer_data_dir = \"/global/scratch/users/aniketh/enformer_data/\"\n",
    "root_save_dir = \"/global/scratch/users/aniketh/finetune-enformer/\"\n",
    "models_dir = \"/global/scratch/users/aniketh/finetune-enformer/saved_models/\"\n",
    "test_preds_dir = \"/global/scratch/users/aniketh/finetune-enformer/test_preds_final/\"\n",
    "rest_unseen_preds_dir = (\n",
    "    \"/global/scratch/users/aniketh/finetune-enformer/rest_unseen_preds_final/\"\n",
    ")\n",
    "ISM_preds_dir = \"/global/scratch/users/aniketh/finetune-enformer/ISM/\"\n",
    "code_dir = \"/global/home/users/aniketh/finetuning-enformer/\"\n",
    "fasta_path = \"/clusterfs/nilah/aniketh/hg19/hg19.fa\"\n",
    "malinois_data_path = (\n",
    "    \"/clusterfs/nilah/aniketh/Malinois/all_sequences_variant_effect_data.csv\"\n",
    ")\n",
    "\n",
    "train_h5_path = os.path.join(data_dir, \"train.h5\")\n",
    "val_h5_path = os.path.join(data_dir, \"val.h5\")\n",
    "test_h5_path = os.path.join(data_dir, \"test.h5\")\n",
    "rest_unseen_h5_path = os.path.join(data_dir, \"rest_unseen.h5\")\n",
    "\n",
    "counts_path = os.path.join(\n",
    "    code_dir, \"process_geuvadis_data\", \"log_tpm\", \"corrected_log_tpm.annot.csv.gz\"\n",
    ")\n",
    "gene_class_path = os.path.join(\n",
    "    code_dir, \"finetuning\", \"data\", \"h5_bins_384_chrom_split\", \"gene_class.csv\"\n",
    ")\n",
    "assert os.path.exists(counts_path) and os.path.exists(gene_class_path)\n",
    "\n",
    "slurm_template = \"slurm_template.sh\"\n",
    "temp_script_path = \"temp_script.sh\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "697ee3c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_main_run_names = {\n",
    "    \"baseline\": \"\",\n",
    "    \"classification\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_parallel_h5_dataset_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"single_regression_counts\": \"NCCL_P2P_DISABLE=1 python finetuning/train_single_counts_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 2 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    #     \"single_regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_single_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 2 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\", ###DOEST PERFORM WELL, NOT USED\n",
    "    \"joint_classification\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"joint_regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"joint_regression_with_Malinois_MPRA\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_MPRA_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {malinois_data_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"single_regression_counts_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_single_counts_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 2 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "    \"regression_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "    \"classification_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_parallel_h5_dataset_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "    # \"joint_classification_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "    # \"joint_regression_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "    # \"joint_regression_with_Malinois_MPRA_random_init\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_MPRA_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {malinois_data_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint --use_random_init\",\n",
    "}\n",
    "\n",
    "all_seeds = [42, 97, 7]\n",
    "subsample_fracs = [0.2, 0.4, 0.6, 0.8]\n",
    "all_afs = (\n",
    "    list(np.arange(0.01, 0.1, 0.01).round(2))\n",
    "    + list(np.arange(0.1, 0.4, 0.05).round(2))\n",
    "    + list(np.arange(0.41, 0.49, 0.01).round(2))\n",
    ")\n",
    "print(all_afs)\n",
    "afs_cache_path = os.path.join(root_save_dir, \"train.h5.afs.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e512bd3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(test_preds_dir):\n",
    "    os.makedirs(test_preds_dir, exist_ok=True)\n",
    "\n",
    "if not os.path.exists(rest_unseen_preds_dir):\n",
    "    os.makedirs(rest_unseen_preds_dir, exist_ok=True)\n",
    "\n",
    "main_cmd = \"NCCL_P2P_DISABLE=1 python finetuning/test_models.py {test_data_path} {cur_test_preds_dir} {model_type} {cur_checkpoints_dir} --use_reverse_complement --create_best_ckpt_copy\"\n",
    "ISM_cmd = \"NCCL_P2P_DISABLE=1 python finetuning/compute_ISM_scores.py {counts_path} {gene_class_path} {fasta_path} {predictions_dir} {model_type} {cur_checkpoints_dir} --use_reverse_complement --gene_name {gene_name}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "260f041d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MAIN TRAIN RUNS\n",
    "for run in all_main_run_names:\n",
    "    if run == \"baseline\":\n",
    "        continue\n",
    "    if \"random_init\" not in run:\n",
    "        continue\n",
    "    for i, seed in enumerate(all_seeds):\n",
    "        run_name = run\n",
    "        cmd = all_main_run_names[run]\n",
    "        cmd = cmd.replace(\"{run_name}\", run_name)\n",
    "        cmd = cmd.replace(\"{data_seed}\", str(seed))\n",
    "        cmd = cmd.replace(\"{train_h5_path}\", train_h5_path)\n",
    "        cmd = cmd.replace(\"{val_h5_path}\", val_h5_path)\n",
    "        cmd = cmd.replace(\"{enformer_data_dir}\", enformer_data_dir)\n",
    "        cmd = cmd.replace(\"{malinois_data_path}\", malinois_data_path)\n",
    "        cmd = cmd.replace(\"{models_dir}\", models_dir)\n",
    "\n",
    "        print(cmd)\n",
    "\n",
    "        temp_script = open(temp_script_path, \"w+\")\n",
    "        for line in open(slurm_template, \"r\").readlines():\n",
    "            temp_script.write(line)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.write(cmd)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.close()\n",
    "\n",
    "        os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "310aed32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SUBSAMPLED TRAIN SET RUNS FOR REGRESSION\n",
    "for run in [\"regression\"]:\n",
    "    for i, seed in enumerate(all_seeds):\n",
    "        for j, frac in enumerate(subsample_fracs):\n",
    "            run_name = run\n",
    "            cmd = all_main_run_names[run]\n",
    "            cmd = cmd.replace(\"{run_name}\", run_name)\n",
    "            cmd = cmd.replace(\"{data_seed}\", str(seed))\n",
    "            cmd = cmd.replace(\"{train_h5_path}\", train_h5_path)\n",
    "            cmd = cmd.replace(\"{val_h5_path}\", val_h5_path)\n",
    "            cmd = cmd.replace(\"{enformer_data_dir}\", enformer_data_dir)\n",
    "            cmd = cmd.replace(\"{models_dir}\", models_dir)\n",
    "            cmd = cmd + f\" --train_set_subsample_ratio {frac}\"\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccd11669-b403-49df-ae5f-567faccd4d1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MAIN TEST RUNS\n",
    "for run in all_main_run_names:\n",
    "    if run == \"baseline\":\n",
    "        model_name = \"baseline\"\n",
    "        cmd = main_cmd.replace(\"{test_data_path}\", test_h5_path)\n",
    "        cmd = cmd.replace(\n",
    "            \"{cur_test_preds_dir}\", os.path.join(test_preds_dir, model_name)\n",
    "        )\n",
    "        cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "        cmd = cmd.replace(\"{cur_checkpoints_dir}\", \"dummy\")\n",
    "\n",
    "        if os.path.exists(os.path.join(test_preds_dir, model_name, \"test_preds.npz\")):\n",
    "            print(f\"{model_name} predictions done\")\n",
    "            continue\n",
    "\n",
    "        print(cmd)\n",
    "\n",
    "        temp_script = open(temp_script_path, \"w+\")\n",
    "        for line in open(slurm_template, \"r\").readlines():\n",
    "            temp_script.write(line)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.write(cmd)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.close()\n",
    "        os.system(f\"sbatch --requeue {temp_script_path}\")\n",
    "    else:\n",
    "        for i, seed in enumerate(all_seeds):\n",
    "\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}\"\n",
    "            if \"random_init\" in run:\n",
    "                model_name = model_name + \"_random_init\"\n",
    "\n",
    "            cmd = main_cmd.replace(\"{test_data_path}\", test_h5_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_test_preds_dir}\", os.path.join(test_preds_dir, model_name)\n",
    "            )\n",
    "            if \"random_init\" in run:\n",
    "                cmd = cmd.replace(\n",
    "                    \"{model_type}\", run[: -len(\"_random_init\")]\n",
    "                )  # run aliases are same as model_type but remove suffix\n",
    "            else:\n",
    "                cmd = cmd.replace(\n",
    "                    \"{model_type}\", run\n",
    "                )  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_checkpoints_dir}\",\n",
    "                os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "            )\n",
    "            assert os.path.exists(os.path.join(models_dir, model_name, \"checkpoints\"))\n",
    "\n",
    "            if os.path.exists(\n",
    "                os.path.join(test_preds_dir, model_name, \"test_preds.npz\")\n",
    "            ):\n",
    "                print(f\"{run} predictions done\")\n",
    "                continue\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            # os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de995b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# REST UNSEEN RUNS\n",
    "cnt = 0\n",
    "for run in reversed(all_main_run_names):\n",
    "    if run == \"baseline\":\n",
    "        model_name = \"baseline\"\n",
    "        cmd = main_cmd.replace(\"{test_data_path}\", rest_unseen_h5_path)\n",
    "        cmd = cmd.replace(\n",
    "            \"{cur_test_preds_dir}\", os.path.join(rest_unseen_preds_dir, model_name)\n",
    "        )\n",
    "        cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "        cmd = cmd.replace(\"{cur_checkpoints_dir}\", \"dummy\")\n",
    "\n",
    "        if os.path.exists(\n",
    "            os.path.join(rest_unseen_preds_dir, model_name, \"test_preds.npz\")\n",
    "        ):\n",
    "            print(f\"{model_name} predictions done\")\n",
    "            continue\n",
    "\n",
    "        print(cmd)\n",
    "\n",
    "        temp_script = open(temp_script_path, \"w+\")\n",
    "        for line in open(slurm_template, \"r\").readlines():\n",
    "            temp_script.write(line)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.write(cmd)\n",
    "        temp_script.write(\"\\n\")\n",
    "\n",
    "        temp_script.close()\n",
    "\n",
    "        os.system(f\"sbatch --requeue {temp_script_path}\")\n",
    "    else:\n",
    "        for i, seed in enumerate(reversed(all_seeds)):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}\"\n",
    "            if \"random_init\" in run:\n",
    "                model_name = model_name + \"_random_init\"\n",
    "\n",
    "            cmd = main_cmd.replace(\"{test_data_path}\", rest_unseen_h5_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_test_preds_dir}\", os.path.join(rest_unseen_preds_dir, model_name)\n",
    "            )\n",
    "            if \"random_init\" in run:\n",
    "                cmd = cmd.replace(\n",
    "                    \"{model_type}\", run[: -len(\"_random_init\")]\n",
    "                )  # run aliases are same as model_type but remove suffix\n",
    "            else:\n",
    "                cmd = cmd.replace(\n",
    "                    \"{model_type}\", run\n",
    "                )  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_checkpoints_dir}\",\n",
    "                os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "            )\n",
    "            assert os.path.exists(os.path.join(models_dir, model_name, \"checkpoints\"))\n",
    "\n",
    "            if os.path.exists(\n",
    "                os.path.join(rest_unseen_preds_dir, model_name, \"test_preds.npz\")\n",
    "            ):\n",
    "                print(f\"{model_name} predictions done\")\n",
    "                continue\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            # os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258149ed-90a7-4e0b-b528-5044c53f4ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST RUNS FOR SUBSAMPLED TRAIN SET RUNS FOR REGRESSION\n",
    "for run in [\"regression\"]:\n",
    "    for i, seed in enumerate(all_seeds):\n",
    "        for j, frac in enumerate(subsample_fracs):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}_subsample_ratio_{frac}\"\n",
    "\n",
    "            cmd = main_cmd.replace(\"{test_data_path}\", test_h5_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_test_preds_dir}\", os.path.join(test_preds_dir, model_name)\n",
    "            )\n",
    "            cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_checkpoints_dir}\",\n",
    "                os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "            )\n",
    "            assert os.path.exists(os.path.join(models_dir, model_name, \"checkpoints\"))\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfcaacbe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# REST UNSEEN RUNS FOR SUBSAMPLED TRAIN SET RUNS FOR REGRESSION\n",
    "for run in [\"regression\"]:\n",
    "    for i, seed in enumerate(all_seeds):\n",
    "        for j, frac in enumerate(subsample_fracs):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}_subsample_ratio_{frac}\"\n",
    "\n",
    "            cmd = main_cmd.replace(\"{test_data_path}\", rest_unseen_h5_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_test_preds_dir}\", os.path.join(rest_unseen_preds_dir, model_name)\n",
    "            )\n",
    "            cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_checkpoints_dir}\",\n",
    "                os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "            )\n",
    "            assert os.path.exists(os.path.join(models_dir, model_name, \"checkpoints\"))\n",
    "\n",
    "            if os.path.exists(\n",
    "                os.path.join(rest_unseen_preds_dir, model_name, \"test_preds.npz\")\n",
    "            ):\n",
    "                print(f\"{model_name} predictions done\")\n",
    "                continue\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "204792d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST RUNS USING SUBSAMPLED VARIANTS FOR REGRESSION\n",
    "for run in [\"regression\"]:\n",
    "    for i, seed in enumerate(all_seeds):\n",
    "        for j, af in enumerate(all_afs):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}\"\n",
    "\n",
    "            cmd = main_cmd.replace(\"{test_data_path}\", test_h5_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_test_preds_dir}\", os.path.join(test_preds_dir, model_name)\n",
    "            )\n",
    "            cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\n",
    "                \"{cur_checkpoints_dir}\",\n",
    "                os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "            )\n",
    "            cmd = (\n",
    "                cmd\n",
    "                + f\" --rare_variant_af_threshold {af} --train_h5_path_for_af_computation {train_h5_path} --afs_cache_path {afs_cache_path}\"\n",
    "            )\n",
    "            assert os.path.exists(os.path.join(models_dir, model_name, \"checkpoints\"))\n",
    "\n",
    "            print(cmd)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "\n",
    "            os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c54eb3-91dd-408f-ad0a-161f8a133120",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Get ISM scores for every gene\n",
    "gene_class_df = pd.read_csv(gene_class_path)\n",
    "population_split_genes = gene_class_df[\n",
    "    gene_class_df[\"class\"] == \"yri_split\"\n",
    "].reset_index(drop=True)\n",
    "print(f\"Number of population-split genes: {len(population_split_genes)}\")\n",
    "\n",
    "cnt = 0\n",
    "\n",
    "for run in [\"baseline\", \"regression\"]:\n",
    "    if run == \"regression\":\n",
    "        for i, seed in enumerate(all_seeds):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}\"\n",
    "\n",
    "            for gene in population_split_genes[\"gene\"].unique():\n",
    "                cmd = ISM_cmd.replace(\"{counts_path}\", counts_path)\n",
    "                cmd = cmd.replace(\"{gene_class_path}\", gene_class_path)\n",
    "                cmd = cmd.replace(\"{fasta_path}\", fasta_path)\n",
    "                cmd = cmd.replace(\n",
    "                    \"{predictions_dir}\", os.path.join(ISM_preds_dir, model_name)\n",
    "                )\n",
    "                cmd = cmd.replace(\n",
    "                    \"{model_type}\", run\n",
    "                )  # run aliases are same as model_type\n",
    "                cmd = cmd.replace(\n",
    "                    \"{cur_checkpoints_dir}\",\n",
    "                    os.path.join(models_dir, model_name, \"checkpoints\"),\n",
    "                )\n",
    "                cmd = cmd.replace(\"{gene_name}\", gene)\n",
    "\n",
    "                temp_script = open(temp_script_path, \"w+\")\n",
    "                for line in open(slurm_template, \"r\").readlines():\n",
    "                    temp_script.write(line)\n",
    "                temp_script.write(\"\\n\")\n",
    "\n",
    "                temp_script.write(cmd)\n",
    "                temp_script.write(\"\\n\")\n",
    "\n",
    "                temp_script.close()\n",
    "                if not os.path.exists(\n",
    "                    os.path.join(ISM_preds_dir, model_name, gene, \"ism_scores.npz\")\n",
    "                ):\n",
    "                    cnt += 1\n",
    "                    print(seed, gene)\n",
    "                    print(cmd)\n",
    "                    os.system(f\"sbatch --requeue {temp_script_path}\")\n",
    "    elif run == \"baseline\":\n",
    "        model_name = \"baseline\"\n",
    "        for gene in population_split_genes[\"gene\"].unique():\n",
    "            cmd = ISM_cmd.replace(\"{counts_path}\", counts_path)\n",
    "            cmd = cmd.replace(\"{gene_class_path}\", gene_class_path)\n",
    "            cmd = cmd.replace(\"{fasta_path}\", fasta_path)\n",
    "            cmd = cmd.replace(\n",
    "                \"{predictions_dir}\", os.path.join(ISM_preds_dir, model_name)\n",
    "            )\n",
    "            cmd = cmd.replace(\"{model_type}\", run)  # run aliases are same as model_type\n",
    "            cmd = cmd.replace(\"{cur_checkpoints_dir}\", \"dummy\")\n",
    "            cmd = cmd.replace(\"{gene_name}\", gene)\n",
    "\n",
    "            temp_script = open(temp_script_path, \"w+\")\n",
    "            for line in open(slurm_template, \"r\").readlines():\n",
    "                temp_script.write(line)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.write(cmd)\n",
    "            temp_script.write(\"\\n\")\n",
    "\n",
    "            temp_script.close()\n",
    "            if not os.path.exists(\n",
    "                os.path.join(ISM_preds_dir, model_name, gene, \"ism_scores.npz\")\n",
    "            ):\n",
    "                cnt += 1\n",
    "                print(gene)\n",
    "                print(cmd)\n",
    "                os.system(f\"sbatch --requeue {temp_script_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4894d327-ca60-4c75-a1d1-d548d65f41db",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "promoter_models_clusterfs",
   "language": "python",
   "name": "promoter_models_clusterfs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
