{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9c4875d0-1acb-4964-b86b-9301d1010dfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pdb\n",
    "import h5py\n",
    "import pickle\n",
    "from collections import defaultdict\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.offsetbox import AnchoredText\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "101d55ad-3a11-40b3-a67d-26cdb4e900da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: <object object at 0x14816106b200>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<contextlib.ExitStack at 0x1480d94e2010>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%matplotlib auto\n",
    "plt.ioff()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0cfd0973-d35b-4649-84ec-243eda71c259",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(3, 2, figsize=(10, 15), tight_layout=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "364c0d42-f724-458b-8b3a-6eca568c2d8d",
   "metadata": {},
   "source": [
    "# Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1533ae8f-3992-42d7-9a4b-68071d9fbc30",
   "metadata": {},
   "outputs": [],
   "source": [
    "geuvadis_eQTL_dir = \"/clusterfs/nilah/Geuvadis/E-GEUV-1/analysis_results\"\n",
    "geuvadis_genotypes_dir = \"/clusterfs/nilah/Geuvadis/E-GEUV-1/genotypes/\"\n",
    "data_dir = \"/clusterfs/nilah/ruchir/src/finetuning-enformer/finetuning/data/h5_bins_384_chrom_split/\"\n",
    "enformer_data_dir = \"/global/scratch/users/aniketh/enformer_data/\"\n",
    "root_save_dir = \"/global/scratch/users/aniketh/finetune-enformer/\"\n",
    "code_dir = \"/global/home/users/aniketh/finetuning-enformer/\"\n",
    "fasta_path = \"/clusterfs/nilah/aniketh/hg19/hg19.fa\"\n",
    "class_path = \"../finetuning/data/h5_bins_384_chrom_split/class.csv\"\n",
    "\n",
    "models_dir = os.path.join(root_save_dir, \"saved_models\")\n",
    "test_preds_dir = os.path.join(root_save_dir, \"test_preds_final\")\n",
    "rest_unseen_preds_dir = os.path.join(root_save_dir, \"rest_unseen_preds_final\")\n",
    "ISM_preds_dir = os.path.join(root_save_dir, \"ISM\")\n",
    "\n",
    "train_h5_path = os.path.join(data_dir, \"train.h5\")\n",
    "val_h5_path = os.path.join(data_dir, \"val.h5\")\n",
    "test_h5_path = os.path.join(data_dir, \"test.h5\")\n",
    "rest_unseen_h5_path = os.path.join(data_dir, \"rest_unseen.h5\")\n",
    "\n",
    "GEUVADIS_COUNTS_PATH = \"../process_geuvadis_data/log_tpm/corrected_log_tpm.annot.csv.gz\"\n",
    "FUSION_PREDS_DIR = \"../fusion/preds\"\n",
    "FUSION_SCALING_PREDS_DIR = \"../fusion/preds_scaling/\"\n",
    "\n",
    "all_main_run_names = {\n",
    "    #     \"classification\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_parallel_h5_dataset_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    #     \"single_regression_counts\": \"NCCL_P2P_DISABLE=1 python finetuning/train_single_counts_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 2 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    # #     \"single_regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_single_parallel_h5_dataset.py {train_h5_path} {val_h5_path} {run_name} {models_dir} --batch_size 2 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\", ###DOEST PERFORM WELL, NOT USED\n",
    "    #     \"joint_classification\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_classification_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    #     \"joint_regression\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_enformer_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {enformer_data_dir} {run_name} {models_dir} --batch_size 1 --lr 0.0005 --weight_decay 0.005 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    #     \"joint_regression_with_Malinois_MPRA\": \"NCCL_P2P_DISABLE=1 python finetuning/train_pairwise_regression_with_MPRA_data_parallel_h5_dynamic_sampling_dataset.py {train_h5_path} {val_h5_path} {malinois_data_path} {run_name} {models_dir} --batch_size 1 --lr 0.0001 --weight_decay 0.001 --use_scheduler --warmup_steps 1000 --data_seed {data_seed} --resume_from_checkpoint\",\n",
    "    \"baseline\": \"\",\n",
    "}\n",
    "all_seeds = [42, 97, 7]\n",
    "subsample_fracs = [0.2, 0.4, 0.6, 0.8]\n",
    "all_afs = (\n",
    "    list(np.arange(0.01, 0.1, 0.01).round(2))\n",
    "    + list(np.arange(0.1, 0.4, 0.05).round(2))\n",
    "    + list(np.arange(0.41, 0.49, 0.01).round(2))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b5a57558-2ab9-41d7-90ac-6ba90ba84f79",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALL_PREDS_PATHS = {}\n",
    "ALL_REST_UNSEEN_PREDS_PATHS = {}\n",
    "\n",
    "# MAIN TEST RUNS\n",
    "for run in all_main_run_names:\n",
    "    if run == \"baseline\":\n",
    "        ALL_PREDS_PATHS[run] = os.path.join(test_preds_dir, run)\n",
    "        ALL_REST_UNSEEN_PREDS_PATHS[run] = os.path.join(rest_unseen_preds_dir, run)\n",
    "    else:\n",
    "        for i, seed in enumerate(all_seeds):\n",
    "            train_cmd_template = all_main_run_names[run]\n",
    "            lr_used_during_training = train_cmd_template.split(\"--lr \")[-1].split(\" \")[\n",
    "                0\n",
    "            ]\n",
    "            wd_used_during_training = train_cmd_template.split(\"--weight_decay \")[\n",
    "                -1\n",
    "            ].split(\" \")[0]\n",
    "            rcprob_used_during_training = 0.5\n",
    "            rsmax_used_during_training = 3\n",
    "\n",
    "            model_name = f\"{run}_data_seed_{seed}_lr_{lr_used_during_training}_wd_{wd_used_during_training}_rcprob_{rcprob_used_during_training}_rsmax_{rsmax_used_during_training}\"\n",
    "            ALL_PREDS_PATHS[model_name] = os.path.join(test_preds_dir, model_name)\n",
    "            ALL_REST_UNSEEN_PREDS_PATHS[model_name] = os.path.join(\n",
    "                rest_unseen_preds_dir, model_name\n",
    "            )\n",
    "\n",
    "test_h5 = h5py.File(test_h5_path, \"r\")\n",
    "rest_unseen_h5 = h5py.File(rest_unseen_h5_path, \"r\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1b26708-c7f2-4623-b297-ae12efe53344",
   "metadata": {},
   "source": [
    "# Read cached correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "208ce2fa-9a32-40f1-addd-178284deea84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from figure_1c+3a+3b+S3+S4+S5.ipynb\n",
    "gene_perf = pd.read_csv(\"all_gene_perf.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc71d1b2-f04f-4868-ae06-d3a86a3d0d75",
   "metadata": {},
   "source": [
    "### Compute the percent of all variants seen during training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "930f051a-4b76-4a43-8a40-2a5a8e697da3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if not os.path.exists(\"gene_variant_stats.csv\"):\n",
    "    train_h5 = h5py.File(train_h5_path, \"r\")\n",
    "    train_genes = train_h5[\"genes\"][:].astype(str)\n",
    "\n",
    "    val_h5 = h5py.File(val_h5_path, \"r\")\n",
    "    val_genes = val_h5[\"genes\"][:].astype(str)\n",
    "\n",
    "    test_genes = test_h5[\"genes\"][:].astype(str)\n",
    "    rest_unseen_genes = rest_unseen_h5[\"genes\"][:].astype(str)\n",
    "\n",
    "    all_percent_seen_during_training = []\n",
    "    all_num_train_variants = []\n",
    "    all_percent_only_seen_during_validation = []\n",
    "    all_num_only_validation = []\n",
    "    all_percent_only_seen_during_testing = []\n",
    "    all_num_only_test = []\n",
    "    all_num_all_variants = []\n",
    "\n",
    "    gene_variant_stats = (\n",
    "        gene_perf[[\"gene\", \"class\"]].drop_duplicates().reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    for i in tqdm(range(len(gene_variant_stats))):\n",
    "        row = gene_variant_stats.iloc[i]\n",
    "        gene_name = row[\"gene\"]\n",
    "        gene_class = row[\"class\"]\n",
    "        print(gene_name, gene_class)\n",
    "        train_gene_seqs = train_h5[\"seqs\"][train_genes == gene_name]\n",
    "        val_gene_seqs = val_h5[\"seqs\"][val_genes == gene_name]\n",
    "        test_gene_seqs = test_h5[\"seqs\"][test_genes == gene_name]\n",
    "        rest_unseen_gene_seqs = rest_unseen_h5[\"seqs\"][rest_unseen_genes == gene_name]\n",
    "\n",
    "        # get variants in each split\n",
    "        train_variants = np.any(train_gene_seqs > 0, axis=(0, 1))  # (seqlen, 4)\n",
    "        val_variants = np.any(val_gene_seqs > 0, axis=(0, 1))\n",
    "        test_variants = (\n",
    "            np.any((test_gene_seqs > 0), axis=(0, 1))\n",
    "            + np.any(rest_unseen_gene_seqs > 0, axis=(0, 1))\n",
    "        ) > 0\n",
    "\n",
    "        # get all variants\n",
    "        all_variants = np.stack([train_variants, val_variants, test_variants])\n",
    "        all_variants = np.any(all_variants, axis=0)\n",
    "        num_all_variants = (all_variants[(all_variants.sum(-1) > 1)].sum(-1) - 1).sum()\n",
    "\n",
    "        # compute percent of variants seen in the training set\n",
    "        # this XOR will find variants that are not present in the train set but are found in the val or test set\n",
    "        if gene_class == \"unseen\":\n",
    "            all_percent_seen_during_training.append(0)\n",
    "            all_num_train_variants.append(0)\n",
    "            all_num_all_variants.append(num_all_variants)\n",
    "            all_percent_only_seen_during_validation.append(0)\n",
    "            all_num_only_test.append(0)\n",
    "            all_percent_only_seen_during_testing.append(100.0)\n",
    "            all_num_only_test.append(num_all_variants)\n",
    "        else:\n",
    "            num_not_seen_in_train = np.logical_xor(all_variants, train_variants).sum()\n",
    "            num_seen_in_train = num_all_variants - num_not_seen_in_train\n",
    "            percent_seen_during_training = (\n",
    "                num_seen_in_train / num_all_variants\n",
    "            ) * 100.0\n",
    "            assert percent_seen_during_training >= 0\n",
    "            print(f\"Num all variants = {num_all_variants}\")\n",
    "            print(\n",
    "                f\"Num variants seen during training = {num_seen_in_train} ({percent_seen_during_training}%)\"\n",
    "            )\n",
    "\n",
    "            all_percent_seen_during_training.append(percent_seen_during_training)\n",
    "            all_num_train_variants.append(num_seen_in_train)\n",
    "            all_num_all_variants.append(num_all_variants)\n",
    "\n",
    "            # compute percent of variants that are only seen in the test set\n",
    "            # this op will find variants that are present in the test set but are not found in the val or test set\n",
    "            num_only_seen_in_test = (\n",
    "                test_variants & np.logical_not(train_variants | val_variants)\n",
    "            ).sum()\n",
    "            num_only_seen_in_validation = (\n",
    "                val_variants & np.logical_not(train_variants | test_variants)\n",
    "            ).sum()\n",
    "\n",
    "            percent_only_seen_during_validation = (\n",
    "                num_only_seen_in_validation / num_all_variants\n",
    "            ) * 100.0\n",
    "            print(\n",
    "                f\"Num variants only seen during validation = {num_only_seen_in_validation} ({percent_only_seen_during_validation}%)\"\n",
    "            )\n",
    "\n",
    "            all_percent_only_seen_during_validation.append(\n",
    "                percent_only_seen_during_validation\n",
    "            )\n",
    "            all_num_only_validation.append(num_only_seen_in_validation)\n",
    "\n",
    "            percent_only_seen_during_testing = (\n",
    "                num_only_seen_in_test / num_all_variants\n",
    "            ) * 100.0\n",
    "            print(\n",
    "                f\"Num variants only seen during testing = {num_only_seen_in_test} ({percent_only_seen_during_testing}%)\"\n",
    "            )\n",
    "\n",
    "            all_percent_only_seen_during_testing.append(\n",
    "                percent_only_seen_during_testing\n",
    "            )\n",
    "            all_num_only_test.append(num_only_seen_in_test)\n",
    "\n",
    "    gene_variant_stats[\n",
    "        \"Percent of all variants seen during training\"\n",
    "    ] = all_percent_seen_during_training\n",
    "    gene_variant_stats[\"Num variants seen during training\"] = all_num_train_variants\n",
    "\n",
    "    gene_variant_stats[\n",
    "        \"Percent of all variants only seen during validation\"\n",
    "    ] = all_percent_only_seen_during_validation\n",
    "    gene_variant_stats[\n",
    "        \"Num variants only seen during validation\"\n",
    "    ] = all_num_only_validation\n",
    "\n",
    "    gene_variant_stats[\n",
    "        \"Percent of all variants only seen during testing\"\n",
    "    ] = all_percent_only_seen_during_testing\n",
    "    gene_variant_stats[\"Num variants only seen during testing\"] = all_num_only_test\n",
    "\n",
    "    gene_variant_stats[\"Num variants\"] = all_num_all_variants\n",
    "\n",
    "    gene_variant_stats.to_csv(\"gene_variant_stats.csv\", index=False)\n",
    "\n",
    "gene_variant_stats = pd.read_csv(\"gene_variant_stats.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de363f2f-2247-4570-b674-d7df9a9e5f0f",
   "metadata": {},
   "source": [
    "## Find strongest eQTL within context window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0bbb84a1-7808-4394-a46c-82d518b62f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(\n",
    "    os.path.join(data_dir, \"YRI_eQTLs_within_context_for_selected_genes.csv\")\n",
    "):\n",
    "    geuvadis_id_to_dbSNP_id = pd.read_csv(\n",
    "        os.path.join(\n",
    "            geuvadis_genotypes_dir, \"Phase1.Geuvadis_dbSnp137_idconvert.txt.gz\"\n",
    "        ),\n",
    "        names=[\"dbSNP_id\", \"geuvadis_id\"],\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "    geuvadis_id_annotations = pd.read_csv(\n",
    "        os.path.join(\n",
    "            geuvadis_genotypes_dir,\n",
    "            \"ALL.phase1_release_v3.20101123.snps_indels_sv.sites.gdid.gdannot.v2.vcf.gz\",\n",
    "        ),\n",
    "        skiprows=78,\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "    geuvadis_id_annotations = geuvadis_id_annotations.merge(\n",
    "        geuvadis_id_to_dbSNP_id, left_on=\"ID\", right_on=\"geuvadis_id\", how=\"inner\"\n",
    "    )\n",
    "\n",
    "    EUR_all_eQTLs_path = os.path.join(\n",
    "        geuvadis_eQTL_dir, \"EUR373.gene.cis.FDR5.all.rs137.txt.gz\"\n",
    "    )\n",
    "    YRI_all_eQTLs_path = os.path.join(\n",
    "        geuvadis_eQTL_dir, \"YRI89.gene.cis.FDR5.all.rs137.txt.gz\"\n",
    "    )\n",
    "\n",
    "    EUR_all_eQTLs = pd.read_csv(EUR_all_eQTLs_path, sep=\"\\t\")\n",
    "    EUR_all_eQTLs = EUR_all_eQTLs[\n",
    "        EUR_all_eQTLs[\"distance\"] < (context_size // 2)\n",
    "    ].reset_index(drop=True)\n",
    "    EUR_all_eQTLs[\"abs_rvalue\"] = np.abs(EUR_all_eQTLs[\"rvalue\"])\n",
    "    EUR_all_eQTLs[\"stable_id\"] = EUR_all_eQTLs.apply(\n",
    "        lambda x: x[\"GENE_ID\"].split(\".\")[0], axis=1\n",
    "    )\n",
    "    EUR_all_eQTLs = EUR_all_eQTLs.sort_values(\n",
    "        by=[\"GENE_ID\", \"abs_rvalue\"], ascending=[True, False]\n",
    "    ).reset_index(drop=True)\n",
    "\n",
    "    YRI_all_eQTLs = pd.read_csv(YRI_all_eQTLs_path, sep=\"\\t\")\n",
    "    YRI_all_eQTLs = YRI_all_eQTLs[\n",
    "        YRI_all_eQTLs[\"distance\"] < (context_size // 2)\n",
    "    ].reset_index(drop=True)\n",
    "    YRI_all_eQTLs[\"abs_rvalue\"] = np.abs(YRI_all_eQTLs[\"rvalue\"])\n",
    "    YRI_all_eQTLs[\"stable_id\"] = YRI_all_eQTLs.apply(\n",
    "        lambda x: x[\"GENE_ID\"].split(\".\")[0], axis=1\n",
    "    )\n",
    "    YRI_all_eQTLs = YRI_all_eQTLs.sort_values(\n",
    "        by=[\"GENE_ID\", \"abs_rvalue\"], ascending=[True, False]\n",
    "    ).reset_index(drop=True)\n",
    "\n",
    "    EUR_all_eQTLs = EUR_all_eQTLs.merge(\n",
    "        geuvadis_id_annotations[\n",
    "            [\"dbSNP_id\", \"geuvadis_id\", \"#CHROM\", \"POS\", \"REF\", \"ALT\", \"INFO\"]\n",
    "        ],\n",
    "        left_on=\"SNP_ID\",\n",
    "        right_on=\"dbSNP_id\",\n",
    "        how=\"inner\",\n",
    "    )\n",
    "    YRI_all_eQTLs = YRI_all_eQTLs.merge(\n",
    "        geuvadis_id_annotations[\n",
    "            [\"dbSNP_id\", \"geuvadis_id\", \"#CHROM\", \"POS\", \"REF\", \"ALT\", \"INFO\"]\n",
    "        ],\n",
    "        left_on=\"SNP_ID\",\n",
    "        right_on=\"dbSNP_id\",\n",
    "        how=\"inner\",\n",
    "    )\n",
    "\n",
    "    GEUVADIS_COUNTS = pd.read_csv(GEUVADIS_COUNTS_PATH)\n",
    "    EUR_all_eQTLs = EUR_all_eQTLs.merge(\n",
    "        GEUVADIS_COUNTS[[\"stable_id\", \"our_gene_name\"]], on=\"stable_id\", how=\"inner\"\n",
    "    )\n",
    "    YRI_all_eQTLs = YRI_all_eQTLs.merge(\n",
    "        GEUVADIS_COUNTS[[\"stable_id\", \"our_gene_name\"]], on=\"stable_id\", how=\"inner\"\n",
    "    )\n",
    "\n",
    "    EUR_eQTLs_within_context_for_selected_genes = gene_variant_stats.merge(\n",
    "        EUR_all_eQTLs, left_on=\"gene\", right_on=\"our_gene_name\", how=\"inner\"\n",
    "    )\n",
    "    EUR_eQTLs_within_context_for_selected_genes = (\n",
    "        EUR_eQTLs_within_context_for_selected_genes[\n",
    "            EUR_eQTLs_within_context_for_selected_genes.columns[8:]\n",
    "        ]\n",
    "    )\n",
    "    EUR_eQTLs_within_context_for_selected_genes.to_csv(\n",
    "        os.path.join(data_dir, \"EUR_eQTLs_within_context_for_selected_genes.csv\"),\n",
    "        index=False,\n",
    "    )\n",
    "\n",
    "    YRI_eQTLs_within_context_for_selected_genes = gene_variant_stats.merge(\n",
    "        YRI_all_eQTLs, left_on=\"gene\", right_on=\"our_gene_name\", how=\"inner\"\n",
    "    )\n",
    "    YRI_eQTLs_within_context_for_selected_genes = (\n",
    "        YRI_eQTLs_within_context_for_selected_genes[\n",
    "            YRI_eQTLs_within_context_for_selected_genes.columns[8:]\n",
    "        ]\n",
    "    )\n",
    "    YRI_eQTLs_within_context_for_selected_genes.to_csv(\n",
    "        os.path.join(data_dir, \"YRI_eQTLs_within_context_for_selected_genes.csv\"),\n",
    "        index=False,\n",
    "    )\n",
    "\n",
    "EUR_eQTLs_within_context_for_selected_genes = pd.read_csv(\n",
    "    os.path.join(data_dir, \"EUR_eQTLs_within_context_for_selected_genes.csv\")\n",
    ")\n",
    "YRI_eQTLs_within_context_for_selected_genes = pd.read_csv(\n",
    "    os.path.join(data_dir, \"YRI_eQTLs_within_context_for_selected_genes.csv\")\n",
    ")\n",
    "\n",
    "top_EUR_eQTL_for_every_gene = (\n",
    "    EUR_eQTLs_within_context_for_selected_genes.groupby(\"our_gene_name\")\n",
    "    .head(1)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "top_YRI_eQTL_for_every_gene = (\n",
    "    YRI_eQTLs_within_context_for_selected_genes.groupby(\"our_gene_name\")\n",
    "    .head(1)\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9bc2bfb7-ecb5-4b01-a8bb-5286ec97957f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 383/383 [00:00<00:00, 14261.15it/s]\n"
     ]
    }
   ],
   "source": [
    "top_eQTL_AFs = []\n",
    "for i in tqdm(range(len(top_EUR_eQTL_for_every_gene))):\n",
    "    row = top_EUR_eQTL_for_every_gene.iloc[i]\n",
    "    af = float(row[\"INFO\"].split(\";EUR_AF=\")[-1].split(\";\")[0])\n",
    "    af = min(af, 1 - af)\n",
    "    top_eQTL_AFs.append(af)\n",
    "\n",
    "top_EUR_eQTL_for_every_gene[\"EUR_AF\"] = top_eQTL_AFs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db42cb0f-48fb-48d7-ad67-47725ebe4705",
   "metadata": {},
   "source": [
    "# Compute mean performance across replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fe3f5008-1777-4796-95f5-06af4bf3ab20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pearson_finetuned</th>\n",
       "      <th>|Pearson|_finetuned</th>\n",
       "      <th>gene</th>\n",
       "      <th>class</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Pearson_baseline</th>\n",
       "      <th>|Pearson|_baseline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.057469</td>\n",
       "      <td>0.074269</td>\n",
       "      <td>a1bg</td>\n",
       "      <td>unseen</td>\n",
       "      <td>19</td>\n",
       "      <td>0.151831</td>\n",
       "      <td>0.151831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.137963</td>\n",
       "      <td>0.178966</td>\n",
       "      <td>a4galt</td>\n",
       "      <td>unseen</td>\n",
       "      <td>22</td>\n",
       "      <td>0.231653</td>\n",
       "      <td>0.231653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.072919</td>\n",
       "      <td>0.072919</td>\n",
       "      <td>aanat</td>\n",
       "      <td>unseen</td>\n",
       "      <td>17</td>\n",
       "      <td>0.029424</td>\n",
       "      <td>0.029424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.082937</td>\n",
       "      <td>0.082937</td>\n",
       "      <td>aasdh</td>\n",
       "      <td>unseen</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.148702</td>\n",
       "      <td>0.148702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.035450</td>\n",
       "      <td>0.058455</td>\n",
       "      <td>abcb6</td>\n",
       "      <td>unseen</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.058654</td>\n",
       "      <td>0.058654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3254</th>\n",
       "      <td>0.080985</td>\n",
       "      <td>0.080985</td>\n",
       "      <td>zscan21</td>\n",
       "      <td>unseen</td>\n",
       "      <td>7</td>\n",
       "      <td>0.063867</td>\n",
       "      <td>0.063867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255</th>\n",
       "      <td>-0.110306</td>\n",
       "      <td>0.212780</td>\n",
       "      <td>zscan23</td>\n",
       "      <td>unseen</td>\n",
       "      <td>6</td>\n",
       "      <td>0.376974</td>\n",
       "      <td>0.376974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3256</th>\n",
       "      <td>0.217700</td>\n",
       "      <td>0.217700</td>\n",
       "      <td>zswim4</td>\n",
       "      <td>unseen</td>\n",
       "      <td>19</td>\n",
       "      <td>0.386495</td>\n",
       "      <td>0.386495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3257</th>\n",
       "      <td>-0.146993</td>\n",
       "      <td>0.146993</td>\n",
       "      <td>zxdc</td>\n",
       "      <td>unseen</td>\n",
       "      <td>3</td>\n",
       "      <td>0.180485</td>\n",
       "      <td>0.180485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3258</th>\n",
       "      <td>0.044851</td>\n",
       "      <td>0.046683</td>\n",
       "      <td>zyg11a</td>\n",
       "      <td>unseen</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.047173</td>\n",
       "      <td>0.047173</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3259 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Pearson_finetuned  |Pearson|_finetuned     gene   class  Chr  \\\n",
       "0             -0.057469             0.074269     a1bg  unseen   19   \n",
       "1              0.137963             0.178966   a4galt  unseen   22   \n",
       "2             -0.072919             0.072919    aanat  unseen   17   \n",
       "3              0.082937             0.082937    aasdh  unseen    4   \n",
       "4             -0.035450             0.058455    abcb6  unseen    2   \n",
       "...                 ...                  ...      ...     ...  ...   \n",
       "3254           0.080985             0.080985  zscan21  unseen    7   \n",
       "3255          -0.110306             0.212780  zscan23  unseen    6   \n",
       "3256           0.217700             0.217700   zswim4  unseen   19   \n",
       "3257          -0.146993             0.146993     zxdc  unseen    3   \n",
       "3258           0.044851             0.046683   zyg11a  unseen    1   \n",
       "\n",
       "      Pearson_baseline  |Pearson|_baseline  \n",
       "0             0.151831            0.151831  \n",
       "1             0.231653            0.231653  \n",
       "2             0.029424            0.029424  \n",
       "3            -0.148702            0.148702  \n",
       "4            -0.058654            0.058654  \n",
       "...                ...                 ...  \n",
       "3254          0.063867            0.063867  \n",
       "3255          0.376974            0.376974  \n",
       "3256          0.386495            0.386495  \n",
       "3257          0.180485            0.180485  \n",
       "3258         -0.047173            0.047173  \n",
       "\n",
       "[3259 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summarized_gene_perf = gene_perf[\n",
    "    (\n",
    "        gene_perf[\"model\"]\n",
    "        == \"regression_data_seed_42_lr_0.0001_wd_0.001_rcprob_0.5_rsmax_3\"\n",
    "    )\n",
    "    | (\n",
    "        gene_perf[\"model\"]\n",
    "        == \"regression_data_seed_97_lr_0.0001_wd_0.001_rcprob_0.5_rsmax_3\"\n",
    "    )\n",
    "    | (\n",
    "        gene_perf[\"model\"]\n",
    "        == \"regression_data_seed_7_lr_0.0001_wd_0.001_rcprob_0.5_rsmax_3\"\n",
    "    )\n",
    "]\n",
    "summarized_gene_perf = (\n",
    "    summarized_gene_perf.drop(\"model\", axis=1).groupby([\"gene\", \"class\", \"Chr\"]).mean()\n",
    ")\n",
    "summarized_gene_perf[\"gene\"] = summarized_gene_perf.index.get_level_values(0)\n",
    "summarized_gene_perf[\"class\"] = summarized_gene_perf.index.get_level_values(1)\n",
    "summarized_gene_perf[\"Chr\"] = summarized_gene_perf.index.get_level_values(2)\n",
    "summarized_gene_perf = summarized_gene_perf.reset_index(drop=True)\n",
    "summarized_gene_perf = summarized_gene_perf.merge(\n",
    "    gene_perf[(gene_perf[\"model\"] == \"baseline\")],\n",
    "    on=[\"gene\", \"class\", \"Chr\"],\n",
    "    suffixes=(\"_finetuned\", \"_baseline\"),\n",
    "    how=\"inner\",\n",
    ")\n",
    "summarized_gene_perf = summarized_gene_perf.reset_index(drop=True).drop(\"model\", axis=1)\n",
    "summarized_gene_perf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bafa8644-f3aa-4328-aaa7-b1e033cf61dc",
   "metadata": {},
   "source": [
    "# Random split genes analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2bfcefe6-d024-4472-9aaf-54cb61b9d9a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_split_gene_perf = (\n",
    "    summarized_gene_perf[summarized_gene_perf[\"class\"] == \"random_split\"]\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a00d6efc-1cdf-4473-8198-6124bb279bf6",
   "metadata": {},
   "source": [
    "## Performance vs. distance of strongest eQTL within context window to TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3912e8c4-df83-4f3a-9d15-8d5f1ee2ea6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "EUR_top_eQTLs_path = os.path.join(\n",
    "    geuvadis_eQTL_dir, \"EUR373.gene.cis.FDR5.best.rs137.txt.gz\"\n",
    ")\n",
    "YRI_top_eQTLs_path = os.path.join(\n",
    "    geuvadis_eQTL_dir, \"YRI89.gene.cis.FDR5.best.rs137.txt.gz\"\n",
    ")\n",
    "\n",
    "EUR_top_eQTLs = pd.read_csv(\n",
    "    EUR_top_eQTLs_path,\n",
    "    sep=\"\\t\",\n",
    "    names=[\n",
    "        \"SNP_ID\",\n",
    "        \"ID\",\n",
    "        \"GENE_ID\",\n",
    "        \"PROBE_ID\",\n",
    "        \"CHR_SNP\",\n",
    "        \"CHR_GENE\",\n",
    "        \"SNPpos\",\n",
    "        \"TSSpos\",\n",
    "        \"Absolute_Distance_from_TSS\",\n",
    "        \"rvalue\",\n",
    "        \"pvalue\",\n",
    "        \"log10pvalue\",\n",
    "    ],\n",
    ")\n",
    "YRI_top_eQTLs = pd.read_csv(\n",
    "    YRI_top_eQTLs_path,\n",
    "    sep=\"\\t\",\n",
    "    names=[\n",
    "        \"SNP_ID\",\n",
    "        \"ID\",\n",
    "        \"GENE_ID\",\n",
    "        \"PROBE_ID\",\n",
    "        \"CHR_SNP\",\n",
    "        \"CHR_GENE\",\n",
    "        \"SNPpos\",\n",
    "        \"TSSpos\",\n",
    "        \"Absolute_Distance_from_TSS\",\n",
    "        \"rvalue\",\n",
    "        \"pvalue\",\n",
    "        \"log10pvalue\",\n",
    "    ],\n",
    ")\n",
    "\n",
    "GEUVADIS_COUNTS = pd.read_csv(GEUVADIS_COUNTS_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "be894b13-52aa-4663-9085-4d592f5a14bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_split_gene_perf = random_split_gene_perf.merge(\n",
    "    top_EUR_eQTL_for_every_gene[\n",
    "        [\n",
    "            \"our_gene_name\",\n",
    "            \"abs_rvalue\",\n",
    "            \"pvalue\",\n",
    "            \"log10pvalue\",\n",
    "            \"distance\",\n",
    "            \"#CHROM\",\n",
    "            \"POS\",\n",
    "            \"REF\",\n",
    "            \"ALT\",\n",
    "            \"EUR_AF\",\n",
    "        ]\n",
    "    ],\n",
    "    left_on=\"gene\",\n",
    "    right_on=\"our_gene_name\",\n",
    "    how=\"inner\",\n",
    ").drop(\"our_gene_name\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4cb56943-66e0-43df-9cad-34cc4aa9180e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Fine-tuned model performance \\non random-split genes vs.\\nthe distance of the top European eQTL to the TSS\\n(Spearman Correlation = -0.073)')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spr = spearmanr(\n",
    "    random_split_gene_perf[\"distance\"], random_split_gene_perf[\"Pearson_finetuned\"]\n",
    ")[0]\n",
    "pr = pearsonr(\n",
    "    random_split_gene_perf[\"distance\"], random_split_gene_perf[\"Pearson_finetuned\"]\n",
    ")[0]\n",
    "\n",
    "sns.regplot(\n",
    "    data=random_split_gene_perf,\n",
    "    x=\"distance\",\n",
    "    y=\"Pearson_finetuned\",\n",
    "    scatter_kws=dict(alpha=0.5),\n",
    "    line_kws=dict(color=\"r\"),\n",
    "    ax=ax[0][0],\n",
    ")\n",
    "ax[0][0].set_xlabel(\n",
    "    \"Absolute distance of top European eQTL \\nwithin context window (24.5kb) to TSS\"\n",
    ")\n",
    "ax[0][0].set_ylabel(\"Fine-tuned Enformer Pearson Correlation\")\n",
    "ax[0][0].set_title(\n",
    "    f\"Fine-tuned model performance \\non random-split genes vs.\\nthe distance of the top European eQTL to the TSS\\n(Spearman Correlation = {spr.round(3)})\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bb95b8-ff26-4dc8-8b11-ddd1e3bd366b",
   "metadata": {},
   "source": [
    "## Performance vs. log10p-val of strongest eQTL within context window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fe4dc797-d994-46de-b9cc-799cda8c2e2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Fine-tuned model performance \\non random-split genes vs.\\nthe p-value of top European eQTL within context window\\n(Spearman Correlation = 0.602)')"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spr = spearmanr(\n",
    "    random_split_gene_perf[\"log10pvalue\"], random_split_gene_perf[\"Pearson_finetuned\"]\n",
    ")[0]\n",
    "pr = pearsonr(\n",
    "    random_split_gene_perf[\"log10pvalue\"], random_split_gene_perf[\"Pearson_finetuned\"]\n",
    ")[0]\n",
    "\n",
    "sns.regplot(\n",
    "    data=random_split_gene_perf,\n",
    "    x=\"log10pvalue\",\n",
    "    y=\"Pearson_finetuned\",\n",
    "    scatter_kws=dict(alpha=0.5),\n",
    "    line_kws=dict(color=\"r\"),\n",
    "    ax=ax[0][1],\n",
    ")\n",
    "ax[0][1].set_xlabel(\n",
    "    \"-log10(p-value) of top European eQTL \\nwithin context window (24.5kb)\"\n",
    ")\n",
    "ax[0][1].set_xscale(\"log\")\n",
    "ax[0][1].set_ylabel(\"Fine-tuned Enformer Pearson Correlation\")\n",
    "ax[0][1].set_title(\n",
    "    f\"Fine-tuned model performance \\non random-split genes vs.\\nthe p-value of top European eQTL within context window\\n(Spearman Correlation = {spr.round(3)})\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eb140b7-296d-48f8-bd16-b9fe0e7cf1eb",
   "metadata": {},
   "source": [
    "## Performance vs. EUR allele frequency of strongest eQTL within context window"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fd1b0a0e-31c2-4cc4-9b83-f14f2ffa593a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# spr = spearmanr(random_split_gene_perf[\"EUR_AF\"],\n",
    "#                 random_split_gene_perf[\"Pearson_finetuned\"])[0]\n",
    "# pr = pearsonr(random_split_gene_perf[\"EUR_AF\"],\n",
    "#               random_split_gene_perf[\"Pearson_finetuned\"])[0]\n",
    "\n",
    "# sns.regplot(data=random_split_gene_perf,\n",
    "#             x=\"EUR_AF\",\n",
    "#             y=\"Pearson_finetuned\")\n",
    "# plt.xlabel(\"EUR minor allele frequency of top European eQTL \\nwithin context window (24.5kb)\")\n",
    "# plt.ylabel(\"Pearson Correlation\")\n",
    "# plt.title(f\"Relationship between fine-tuned model performance \\non random-split genes \\nand the EUR minor AF of top European eQTL within context window\\n(Spearman Correlation = {spr.round(3)})\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b214aeac-166e-4f27-a13c-219fefbcfbe3",
   "metadata": {},
   "source": [
    "## Performance vs. percentage of variants observed during training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fa6879b4-c20c-4efd-ae33-94b9ae0dfc13",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_split_gene_perf = random_split_gene_perf.merge(\n",
    "    gene_variant_stats[[\"gene\"] + gene_variant_stats.columns[-7:].tolist()],\n",
    "    on=\"gene\",\n",
    "    how=\"inner\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8edc168-5a8d-48d5-88a7-b2dbf9016f01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Random-split genes\\n(Spearman Correlation = 0.112)')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spr = spearmanr(\n",
    "    random_split_gene_perf[\"Percent of all variants seen during training\"],\n",
    "    random_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "pr = pearsonr(\n",
    "    random_split_gene_perf[\"Percent of all variants seen during training\"],\n",
    "    random_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "\n",
    "sns.regplot(\n",
    "    data=random_split_gene_perf,\n",
    "    x=\"Percent of all variants seen during training\",\n",
    "    y=\"Pearson_finetuned\",\n",
    "    scatter_kws=dict(alpha=0.5),\n",
    "    line_kws=dict(color=\"r\"),\n",
    "    ax=ax[1][0],\n",
    ")\n",
    "\n",
    "ax[1][0].set_ylabel(\"Fine-tuned Enformer Pearson Correlation\")\n",
    "ax[1][0].set_title(f\"Random-split genes\\n(Spearman Correlation = {spr.round(3)})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c94cd0f4-c778-4d35-9869-827b073e6582",
   "metadata": {},
   "source": [
    "# Population split genes analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6a8525b5-157b-4436-a48f-9da615771dea",
   "metadata": {},
   "outputs": [],
   "source": [
    "yri_split_gene_perf = (\n",
    "    summarized_gene_perf[summarized_gene_perf[\"class\"] == \"yri_split\"]\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b928514e-3947-4107-a9cc-869c3646bc9f",
   "metadata": {},
   "source": [
    "## Performance vs. percentage of variants observed during training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5e188e4e-586d-42c6-9980-d576195ada28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pearson_finetuned</th>\n",
       "      <th>|Pearson|_finetuned</th>\n",
       "      <th>gene</th>\n",
       "      <th>class</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Pearson_baseline</th>\n",
       "      <th>|Pearson|_baseline</th>\n",
       "      <th>Percent of all variants seen during training</th>\n",
       "      <th>Num variants seen during training</th>\n",
       "      <th>Percent of all variants only seen during testing</th>\n",
       "      <th>Num variants only seen during testing</th>\n",
       "      <th>Num variants</th>\n",
       "      <th>Percent of all variants only seen during validation</th>\n",
       "      <th>Num variants only seen during validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.103805</td>\n",
       "      <td>0.103805</td>\n",
       "      <td>abhd10</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>3</td>\n",
       "      <td>0.042268</td>\n",
       "      <td>0.042268</td>\n",
       "      <td>70.168067</td>\n",
       "      <td>334</td>\n",
       "      <td>29.831933</td>\n",
       "      <td>142</td>\n",
       "      <td>476</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.242679</td>\n",
       "      <td>0.242679</td>\n",
       "      <td>ac017104.1</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>2</td>\n",
       "      <td>0.159504</td>\n",
       "      <td>0.159504</td>\n",
       "      <td>55.803571</td>\n",
       "      <td>250</td>\n",
       "      <td>43.526786</td>\n",
       "      <td>195</td>\n",
       "      <td>448</td>\n",
       "      <td>0.446429</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.382086</td>\n",
       "      <td>0.382086</td>\n",
       "      <td>ac079921.2</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>4</td>\n",
       "      <td>-0.132472</td>\n",
       "      <td>0.132472</td>\n",
       "      <td>57.739558</td>\n",
       "      <td>235</td>\n",
       "      <td>41.523342</td>\n",
       "      <td>169</td>\n",
       "      <td>407</td>\n",
       "      <td>0.737101</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.079437</td>\n",
       "      <td>0.079437</td>\n",
       "      <td>ac104653.1</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.119865</td>\n",
       "      <td>0.119865</td>\n",
       "      <td>74.492099</td>\n",
       "      <td>330</td>\n",
       "      <td>25.507901</td>\n",
       "      <td>113</td>\n",
       "      <td>443</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.476652</td>\n",
       "      <td>0.476652</td>\n",
       "      <td>ac108206.1</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>4</td>\n",
       "      <td>0.508323</td>\n",
       "      <td>0.508323</td>\n",
       "      <td>61.111111</td>\n",
       "      <td>341</td>\n",
       "      <td>38.530466</td>\n",
       "      <td>215</td>\n",
       "      <td>558</td>\n",
       "      <td>0.358423</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>0.229185</td>\n",
       "      <td>0.229185</td>\n",
       "      <td>znf22-as1</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>10</td>\n",
       "      <td>-0.054358</td>\n",
       "      <td>0.054358</td>\n",
       "      <td>58.503401</td>\n",
       "      <td>258</td>\n",
       "      <td>40.589569</td>\n",
       "      <td>179</td>\n",
       "      <td>441</td>\n",
       "      <td>0.907029</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>0.098031</td>\n",
       "      <td>0.098031</td>\n",
       "      <td>znf354a</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>5</td>\n",
       "      <td>0.208737</td>\n",
       "      <td>0.208737</td>\n",
       "      <td>56.060606</td>\n",
       "      <td>296</td>\n",
       "      <td>43.560606</td>\n",
       "      <td>230</td>\n",
       "      <td>528</td>\n",
       "      <td>0.189394</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>0.288661</td>\n",
       "      <td>0.288661</td>\n",
       "      <td>znf354b</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>5</td>\n",
       "      <td>0.271102</td>\n",
       "      <td>0.271102</td>\n",
       "      <td>73.103448</td>\n",
       "      <td>530</td>\n",
       "      <td>26.758621</td>\n",
       "      <td>194</td>\n",
       "      <td>725</td>\n",
       "      <td>0.137931</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>0.262740</td>\n",
       "      <td>0.262740</td>\n",
       "      <td>znf438</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>10</td>\n",
       "      <td>0.353091</td>\n",
       "      <td>0.353091</td>\n",
       "      <td>56.103286</td>\n",
       "      <td>239</td>\n",
       "      <td>43.427230</td>\n",
       "      <td>185</td>\n",
       "      <td>426</td>\n",
       "      <td>0.469484</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>0.162193</td>\n",
       "      <td>0.162193</td>\n",
       "      <td>zp3</td>\n",
       "      <td>yri_split</td>\n",
       "      <td>7</td>\n",
       "      <td>-0.019802</td>\n",
       "      <td>0.019802</td>\n",
       "      <td>70.841487</td>\n",
       "      <td>362</td>\n",
       "      <td>28.767123</td>\n",
       "      <td>147</td>\n",
       "      <td>511</td>\n",
       "      <td>0.391389</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Pearson_finetuned  |Pearson|_finetuned        gene      class  Chr  \\\n",
       "0             0.103805             0.103805      abhd10  yri_split    3   \n",
       "1             0.242679             0.242679  ac017104.1  yri_split    2   \n",
       "2             0.382086             0.382086  ac079921.2  yri_split    4   \n",
       "3             0.079437             0.079437  ac104653.1  yri_split    2   \n",
       "4             0.476652             0.476652  ac108206.1  yri_split    4   \n",
       "..                 ...                  ...         ...        ...  ...   \n",
       "195           0.229185             0.229185   znf22-as1  yri_split   10   \n",
       "196           0.098031             0.098031     znf354a  yri_split    5   \n",
       "197           0.288661             0.288661     znf354b  yri_split    5   \n",
       "198           0.262740             0.262740      znf438  yri_split   10   \n",
       "199           0.162193             0.162193         zp3  yri_split    7   \n",
       "\n",
       "     Pearson_baseline  |Pearson|_baseline  \\\n",
       "0            0.042268            0.042268   \n",
       "1            0.159504            0.159504   \n",
       "2           -0.132472            0.132472   \n",
       "3           -0.119865            0.119865   \n",
       "4            0.508323            0.508323   \n",
       "..                ...                 ...   \n",
       "195         -0.054358            0.054358   \n",
       "196          0.208737            0.208737   \n",
       "197          0.271102            0.271102   \n",
       "198          0.353091            0.353091   \n",
       "199         -0.019802            0.019802   \n",
       "\n",
       "     Percent of all variants seen during training  \\\n",
       "0                                       70.168067   \n",
       "1                                       55.803571   \n",
       "2                                       57.739558   \n",
       "3                                       74.492099   \n",
       "4                                       61.111111   \n",
       "..                                            ...   \n",
       "195                                     58.503401   \n",
       "196                                     56.060606   \n",
       "197                                     73.103448   \n",
       "198                                     56.103286   \n",
       "199                                     70.841487   \n",
       "\n",
       "     Num variants seen during training  \\\n",
       "0                                  334   \n",
       "1                                  250   \n",
       "2                                  235   \n",
       "3                                  330   \n",
       "4                                  341   \n",
       "..                                 ...   \n",
       "195                                258   \n",
       "196                                296   \n",
       "197                                530   \n",
       "198                                239   \n",
       "199                                362   \n",
       "\n",
       "     Percent of all variants only seen during testing  \\\n",
       "0                                           29.831933   \n",
       "1                                           43.526786   \n",
       "2                                           41.523342   \n",
       "3                                           25.507901   \n",
       "4                                           38.530466   \n",
       "..                                                ...   \n",
       "195                                         40.589569   \n",
       "196                                         43.560606   \n",
       "197                                         26.758621   \n",
       "198                                         43.427230   \n",
       "199                                         28.767123   \n",
       "\n",
       "     Num variants only seen during testing  Num variants  \\\n",
       "0                                      142           476   \n",
       "1                                      195           448   \n",
       "2                                      169           407   \n",
       "3                                      113           443   \n",
       "4                                      215           558   \n",
       "..                                     ...           ...   \n",
       "195                                    179           441   \n",
       "196                                    230           528   \n",
       "197                                    194           725   \n",
       "198                                    185           426   \n",
       "199                                    147           511   \n",
       "\n",
       "     Percent of all variants only seen during validation  \\\n",
       "0                                             0.000000     \n",
       "1                                             0.446429     \n",
       "2                                             0.737101     \n",
       "3                                             0.000000     \n",
       "4                                             0.358423     \n",
       "..                                                 ...     \n",
       "195                                           0.907029     \n",
       "196                                           0.189394     \n",
       "197                                           0.137931     \n",
       "198                                           0.469484     \n",
       "199                                           0.391389     \n",
       "\n",
       "     Num variants only seen during validation  \n",
       "0                                           0  \n",
       "1                                           2  \n",
       "2                                           3  \n",
       "3                                           0  \n",
       "4                                           2  \n",
       "..                                        ...  \n",
       "195                                         4  \n",
       "196                                         1  \n",
       "197                                         1  \n",
       "198                                         2  \n",
       "199                                         2  \n",
       "\n",
       "[200 rows x 14 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yri_split_gene_perf = yri_split_gene_perf.merge(\n",
    "    gene_variant_stats[[\"gene\"] + gene_variant_stats.columns[-7:].tolist()],\n",
    "    on=\"gene\",\n",
    "    how=\"inner\",\n",
    ")\n",
    "yri_split_gene_perf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8d894aad-4dc4-4a23-8a36-4b92eb9b90e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Population-split genes\\n(Spearman Correlation = 0.102)')"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spr = spearmanr(\n",
    "    yri_split_gene_perf[\"Percent of all variants seen during training\"],\n",
    "    yri_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "pr = pearsonr(\n",
    "    yri_split_gene_perf[\"Percent of all variants seen during training\"],\n",
    "    yri_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "\n",
    "sns.regplot(\n",
    "    data=yri_split_gene_perf,\n",
    "    x=\"Percent of all variants seen during training\",\n",
    "    y=\"Pearson_finetuned\",\n",
    "    scatter_kws=dict(alpha=0.5),\n",
    "    line_kws=dict(color=\"r\"),\n",
    "    ax=ax[1][1],\n",
    ")\n",
    "ax[1][1].set_ylabel(\"Fine-tuned Enformer Pearson Correlation\")\n",
    "ax[1][1].set_title(f\"Population-split genes\\n(Spearman Correlation = {spr.round(3)})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c80ac277-4f94-42bd-a814-797d8ae10aaa",
   "metadata": {},
   "source": [
    "# Unseen genes analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "51b9ddce-30ef-4bb4-a387-cb749ad8dd61",
   "metadata": {},
   "outputs": [],
   "source": [
    "unseen_split_gene_perf = (\n",
    "    summarized_gene_perf[summarized_gene_perf[\"class\"] == \"unseen\"]\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de314e2e-d4d3-4ce1-a9ff-95d82befd6f0",
   "metadata": {},
   "source": [
    "## Performance vs. Baseline performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3284cfcc-7be4-472e-816b-9629b8c2dcaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, 'Fine-tuned Enformer Pearson Correlation')"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spr = spearmanr(\n",
    "    unseen_split_gene_perf[\"Pearson_baseline\"],\n",
    "    unseen_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "pr = pearsonr(\n",
    "    unseen_split_gene_perf[\"Pearson_baseline\"],\n",
    "    unseen_split_gene_perf[\"Pearson_finetuned\"],\n",
    ")[0]\n",
    "\n",
    "# sns.regplot(data=unseen_split_gene_perf,\n",
    "#             x=\"Pearson_baseline\",\n",
    "#             y=\"Pearson_finetuned\",\n",
    "#             scatter_kws=dict(alpha=0.5),\n",
    "#             line_kws=dict(color=\"r\"),\n",
    "#             ax=ax[2][0])\n",
    "\n",
    "sns.scatterplot(\n",
    "    data=unseen_split_gene_perf,\n",
    "    x=\"Pearson_baseline\",\n",
    "    y=\"Pearson_finetuned\",\n",
    "    alpha=0.5,\n",
    "    ax=ax[2][0],\n",
    ")\n",
    "ax[2][0].set_xlabel(\"Baseline Enformer Pearson Correlation\")\n",
    "ax[2][0].set_ylabel(\"Fine-tuned Enformer Pearson Correlation\")\n",
    "# ax[2][0].set_title(f\"Fine-tuned model performance vs.\\nbaseline Enformer performance \\non unseen genes\\n(Spearman Correlation = {spr.round(3)})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7e22dcc2-4fb9-43d1-a80c-18b0ea61b9c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax[2][1].set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "98001cac-a7fa-4968-b26c-ffc6d4fdb9d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.savefig(\n",
    "    \"figures/fig_S6+S7_correlations_with_gene_features.pdf\",\n",
    "    dpi=600,\n",
    "    bbox_inches=\"tight\",\n",
    ")\n",
    "plt.savefig(\n",
    "    \"figures/fig_S6+S7_correlations_with_gene_features.svg\",\n",
    "    dpi=600,\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "956d8786-87d3-477a-ae60-5ee32ed33758",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "promoter_models_clusterfs",
   "language": "python",
   "name": "promoter_models_clusterfs"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
